posts/inside-rust/2023-11-10-parallel-rustc.md

```diff
@@ -0,0 +1,252 @@
+---
+layout: post
+title: Faster compilation with the parallel front-end in nightly
+author: Nicholas Nethercote
+team: The Parallel Rustc Working Group <https://www.rust-lang.org/governance/teams/compiler#Parallel%20rustc%20working%20group>
+---
+
+The Rust compiler's front-end can now use parallel execution to significantly
+reduce compile times. To try it, run the nightly compiler with the `-Z
+threads=8` option. This feature is currently experimental, and we aim to ship
+it in the stable compiler in 2024.
+
+Keep reading to learn why a parallel front-end is needed and how it works, or
+just skip ahead to the [How to use it](parallel-rustc.html#how-to-use-it)
+section.
+
+## Compile times and parallelism
+
+Rust compile times are a perennial concern. The [Compiler Performance Working
+Group](https://www.rust-lang.org/governance/teams/compiler#Compiler%20performance%20working%20group)
+has continually improved compiler performance for several years. For example,
+in the first 10 months of 2023, there were mean reductions in compile time of
+[13%](https://perf.rust-lang.org/compare.html?start=2023-01-01&end=2023-10-31&stat=wall-time&nonRelevant=true),
+in peak memory use of
+[15%](https://perf.rust-lang.org/compare.html?start=2023-01-01&end=2023-10-31&stat=max-rss&nonRelevant=true),
+and in binary size of
+[7%](https://perf.rust-lang.org/compare.html?start=2023-01-01&end=2023-10-31&stat=size%3Alinked_artifact&nonRelevant=true),
+as measured by our performance suite.
+
+However, at this point the compiler has been heavily optimized and new
+improvements are hard to find. There is no low-hanging fruit remaining. 
+
+But there is one piece of large but high-hanging fruit: parallelism. Current
+Rust compiler users benefit from two kinds of parallelism, and the newly
+parallel front-end adds a third kind.
+
+### Existing interprocess parallelism
+
+When you compile a Rust program, Cargo launches multiple rustc processes,
+compiling multiple crates in parallel. This works well. Try compiling a large
+Rust program with the `-j1` flag to disable this parallelization and it will
+take a lot longer than normal.
```

Comment:
do we need to say here that `-j1` would not just "disable this parallelization" but would also disable the parallelism in rustc's backend?

Reply:
Is that true? Either way, `-j1` will be slower and I think that's an unimportant detail.

Reply:
Maybe `to disable parallelization`?

Reply:
I agree it's not particularly important because the point is that we do have parallelism already and it speeds things up considerably, but it is true yes. (Building a leaf crate at j1 will show the difference, IIRC `syn` is particularly helped by rustc's intraprocess parallelism today.)

Reply:
@Mark-Simulacrum Happy to have this resolved by you or @nnethercote as the distinction is not absolutely needed, but can't mark it resolved myself.